c2c: deflake c2c/shutdown roachtests

msbutler opened the pull request:
c2c: deflake c2c/shutdown roachtests

    This patch addresses to roachtest failure modes:
    - Prevents roachtest failure if a query fails during a node shutdown.

    - Prevents the src cluster from returning a single node topology, which could
      cause the stream ingestion job to hang if the participating src node gets
    shut down. Longer term, automatic replanning will prevent this. Specifically,
    this patch changes the kv workload to split and scatter the kv table across the
    cluster before the c2c job begins.

    Fixes #101898
    Fixes #102111

    This patch also makes it easier to reproduce c2c roachtest failures by plumbing
    a random seed to several components of the roachtest driver.

    Release note: None


    c2c: rename completeStreamIngestion to applyCutoverTime

    Release note: none


    workload: add --scatter flag to kv workload

    The user can now run `./workload init kv --scatter ....` which scatters the kv
    table across the cluster after the initial data load. This flag is best used
    with `--splits`, `--max-block-bytes`, and `--insert-count`.

    Epic: none

    Release note: none

msbutler committed:
c2c: deflake c2c/shutdown roachtests

This patch addresses to roachtest failure modes:
- Prevents roachtest failure if a query fails during a node shutdown.

- Prevents the src cluster from returning a single node topology, which could
  cause the stream ingestion job to hang if the participating src node gets
shut down. Longer term, automatic replanning will prevent this. Specifically,
this patch changes the kv workload to split and scatter the kv table across the
cluster before the c2c job begins.

Fixes #101898
Fixes #102111

This patch also makes it easier to reproduce c2c roachtest failures by plumbing
a random seed to several components of the roachtest driver.

Release note: None

lidorcarmel commented on pkg/cmd/roachtest/tests/cluster_to_cluster.go:
```diff
@@ -963,7 +982,10 @@ func (rrd *replResilienceDriver) waitForTargetPhase() error {
 			// Every C2C phase should last at least 30 seconds, so introduce a little
 			// bit of random waiting before node shutdown to ensure the shutdown occurs
 			// once we're settled into the target phase.
-			randomSleep := time.Duration(rand.Intn(6))
+			//
+			// Sleep for a minimum of 5 seconds to ensure the stream processors have
+			// observed the cutover signal.
+			randomSleep := time.Duration(5 + rand.Intn(6))
 			rrd.t.L().Printf("In target phase! Take a %d second power nap", randomSleep)
 			time.Sleep(randomSleep * time.Second)
```
I know my comment isn't really about this pr but should `waitForTargetPhase` do what it says, and just wait for the target phase without any sleep? and then the caller (node shutdown or something else) do the sleep instead?
when reading this code it's not so clear why node shutdown is mentioned in the comment.
WDYT?

lidorcarmel commented on pkg/cmd/roachtest/tests/cluster_to_cluster.go:
```diff
@@ -617,10 +637,9 @@ func (rd *replicationDriver) main(ctx context.Context) {
 	removeTenantRateLimiters(rd.t, rd.setup.dst.sysSQL, rd.setup.dst.name)
 
-	lv := makeLatencyVerifier("stream-ingestion", 0, 2*time.Minute, rd.t.L(), getStreamIngestionJobInfo, rd.t.Status, false)
+	lv := makeLatencyVerifier("stream-ingestion", 0, 2*time.Minute, rd.t.L(), getStreamIngestionJobInfo, rd.t.Status, true)
```
why do we want to ignore errors here?
don't we want the test to fail if we can't read from `crdb_internal.system_jobs`?

lidorcarmel reviewed (commented):
looks good, a couple of small things..

msbutler committed:
c2c: rename completeStreamIngestion to applyCutoverTime

Release note: none

msbutler committed:
workload: add --scatter flag to kv workload

The user can now run `./workload init kv --scatter ....` which scatters the kv
table across the cluster after the initial data load. This flag is best used
with `--splits`, `--max-block-bytes`, and `--insert-count`.

Epic: none

Release note: none

stevendanna reviewed (approved):
I wonder if it would also help to explicitly scatter the ranges.

renatolabs reviewed (commented):
All ignorable nits, don't mind me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-:-NV6GrjK7j21YCwiIzPH:b-h4kk6b)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel and @msbutler)

___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FCAZ25xKnxPLX4gL:-NV6FCA_67luzv0Tz3pX:babitws) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L243)):*
> ```Go
> 
> func (kv replicateKV) sourceInitCmd(tenantName string, nodes option.NodeListOption) string {
> 	cmd := roachtestutil.NewCommand(`./workload init kv`)
> ```

Nit: you can chain all of these calls which arguably makes it nicer to read:

```go
return roachtestutil.NewCommand("./workload init kv").
    Flag("foo", 1).
    Flag("bar", 2).
    String()
```
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 245 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FhN8AaFoolvtRaqC:-NV6FhN97FLrUB7Dwp20:b-y3ujxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L245)):*
> ```Go
> 	cmd := roachtestutil.NewCommand(`./workload init kv`)
> 	cmd.MaybeFlag(kv.initRows > 0, "insert-count", kv.initRows)
> 	cmd.MaybeFlag(kv.initRows > 0, "max-block-bytes", kv.maxBlockBytes)
> ```

Did you want to check for `maxBlockBytes` here?
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 248 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FSsNC4AfYiqFOKA8:-NV6FSsNC4AfYiqFOKA9:b-i0d69m) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L248)):*
> ```Go
> 	cmd.Flag("splits", 100)
> 	cmd.Option("scatter")
> 	cmd.Arg(fmt.Sprintf("{pgurl%s:%s}", nodes, tenantName))
> ```

Nit: `Arg` is for single argument, and it already understands `Printf`-style formatting, so you can just do `Arg("{pgurl%s:%s}", nodes, tenantName)`
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 1006 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6G0r9AmhU0-Au8w7G:-NV6G0r9AmhU0-Au8w7H:b-77s3ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L1006)):*
> ```Go
> 	// once we're fully settled into the target phase (e.g. the stream ingestion
> 	// processors have observed the cutover signal).
> 	randomSleep := time.Duration(5 + rrd.rng.Intn(6))
> ```

Nit: making `randomSleep` a `time.Duration` (i.e., number of nanoseconds), even though the actual intended meaning is for it to be a number of seconds is slightly confusing. Suggestion:

```go
dur := (5 + rrd.rng.Intn(6)) * time.Second
t.L().Printf("taking a %s power nap", dur)
time.Sleep(dur)
```


<!-- Sent from Reviewable.io -->

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102382)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lidorcarmel and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FCAZ25xKnxPLX4gL:-NVAho_Y1HkqPVG-JW0V:b-mvxkxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L243)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: you can chain all of these calls which arguably makes it nicer to read:

```go
return roachtestutil.NewCommand("./workload init kv").
    Flag("foo", 1).
    Flag("bar", 2).
    String()
```
</blockquote></details>

TIL. Done.
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 245 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FhN8AaFoolvtRaqC:-NVAhqYE4O9EqPsHA0LY:b-iis3ls) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L245)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Did you want to check for `maxBlockBytes` here?
</blockquote></details>

Yeah I did this bc the parameter only seems relevant if i actually insert data, and bc I distrust the workload package's parameter handling logic :D
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 248 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6FSsNC4AfYiqFOKA8:-NVAi8FeCSDppf7Bm4z3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L248)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: `Arg` is for single argument, and it already understands `Printf`-style formatting, so you can just do `Arg("{pgurl%s:%s}", nodes, tenantName)`
</blockquote></details>

Done
___
*[`pkg/cmd/roachtest/tests/cluster_to_cluster.go` line 1006 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/102382#-NV6G0r9AmhU0-Au8w7G:-NVAiATl8eX4yRLF5-k8:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bedd7fcb994b96b82d814ea52aa70f57613a340d/pkg/cmd/roachtest/tests/cluster_to_cluster.go#L1006)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: making `randomSleep` a `time.Duration` (i.e., number of nanoseconds), even though the actual intended meaning is for it to be a number of seconds is slightly confusing. Suggestion:

```go
dur := (5 + rrd.rng.Intn(6)) * time.Second
t.L().Printf("taking a %s power nap", dur)
time.Sleep(dur)
```
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

msbutler commented:
thanks for the nits @renatolabs !

msbutler commented:
TFTRs!!

bors r=stevendanna
